report/sections/TestRunner/chapter.tex

```diff
@@ -1,33 +1,32 @@
-Furthermore, it was deemed easier to implement a shared state among endpoints by leveraging axum's \texttt{layer} feature.
-We give an overview and a more technical explanation of how this was done in section \ref{sec:queue-system}.
-
-When the Test Runner receives the request from Next.js, its singular goal is to run the tests against the code and send the result of this back as fast as possible.
-In order to run the tests, we need to run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
+When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and send back the result as fast as possible.
+In order to run the tests, we run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
 The GHC interpreter can be run from the command line with arguments specifying the path Haskell file to run.
-In other words, the Test Runner needs to write the user submission and corresponding test code to a file.
-However, if more users were to make a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
-
+In other words, the Test Runner writes the user submission and corresponding test code to a file.
+However, if multiple users were to post a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
 To address this, we generate a new directory with a unique name for each submission.
 Then, the user submission is written to two  \texttt{.hs} files inside this directory - a \texttt{code.hs} file containing the exercise submission, and a \texttt{test.hs} file containing the corresponding tests.
 This way, we can pass the path to this isolated directory to the GHC interpreter process as an argument.
 Once the tests have been run, the directory and its contents are deleted again.
 The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
 Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.
 
-Sequentially starting a new \texttt{runhaskell} process for each request was identified as a major bottleneck during testing, which will be described in detail in chapter \ref{chap:Benchmarking}.
-In order to address this, we started by refactoring the code base to allow running several instances of the GHC interpreter process at the same time.
-However, despite this running faster for a small number of clients, this approach in itself is not sufficiently stable:
-If multiple clients were to post a code submission at the same time, the Test Runner will start a new process for each request.
+Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
```

Comment:
```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when users submit solutions in quick succession.
```

Reply:
The idea with "multiple" was to say that it was many users at the same time. Maybe we should just change "multiple" to many instead of removing it?

Reply:
So it is to emphasize that it is _many_ users, not just some number of users? If so, sure, we can change multiple to another word.

Reply:
```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when many multiple users submit solutions in quick succession.
```